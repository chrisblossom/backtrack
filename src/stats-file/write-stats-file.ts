import path from 'path';
import sortKeys from 'sort-keys';
import { existsSync, writeFileSync } from 'fs';
import { isEqual } from 'lodash';
import del from 'del';
import { normalizeStats } from './normalize-stats';
import { rootPath } from '../config/paths';
import log from '../utils/log';

import { StatsFile } from '../types';

const statsFilename = '.backtrack-stats.json';

async function writeStatsFile(
	sections: StatsFile = {},
	previousStats?: StatsFile,
) {
	const statsFile = path.resolve(rootPath, statsFilename);

	const filterEmptySections: StatsFile = Object.keys(sections).reduce(
		(acc, section) => {
			// @ts-ignore
			const matchedSection = sections[section] || {};

			if (Object.keys(matchedSection).length === 0) {
				return acc;
			}

			return {
				...acc,
				[section]: matchedSection,
			};
		},
		{},
	);

	/**
	 * remove stats file if there are no stats
	 */
	const shouldDelete = Object.keys(filterEmptySections).reduce(
		(acc: number[], section) => {
			// @ts-ignore
			const matchedSection = filterEmptySections[section];
			if (matchedSection === undefined) {
				return acc;
			}

			return [...acc, Object.keys(matchedSection).length];
		},
		[],
	);

	if (shouldDelete.length === 0) {
		const exists = existsSync(statsFile);

		/**
		 * Do not remove anything if the file does not exist
		 */
		if (exists === true) {
			await del(statsFile);

			log.info(`No stats found, ${statsFilename} deleted`);
		}

		return;
	}

	const normalized = normalizeStats.write(filterEmptySections);

	/**
	 * Sort stats file to keep consistency for git changes
	 */
	const sorted = sortKeys(normalized, { deep: true });

	/**
	 * Do not write file if stats have not changed
	 */
	if (isEqual(sorted, previousStats)) {
		return;
	}

	writeFileSync(
		statsFile,
		// eslint-disable-next-line prefer-template
		JSON.stringify(sorted, null, 4) + '\n',
	);

	log.info(`${statsFilename} updated`);
}

export { writeStatsFile };
