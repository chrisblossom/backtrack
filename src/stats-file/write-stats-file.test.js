/* @noflow */

import path from 'path';
import fs from 'fs';

require('./write-stats-file');

const writeStatsFile = (stats, previousStats) =>
    require('./write-stats-file').writeStatsFile(stats, previousStats);

describe('writeStatsFile', () => {
    let writeFileSync;
    const cwd = process.cwd();
    let del;

    beforeEach(() => {
        // $FlowIssue
        del = require.requireMock('del');

        jest.mock('del', () => jest.fn(() => Promise.resolve()));

        jest.mock('../utils/log', () => ({
            warn: jest.fn(),
            info: jest.fn(),
            error: jest.fn(),
            success: jest.fn(),
        }));

        /**
         * Before calling '.toMatchSnapshot();' snapshot, call 'writeFileSync.mockRestore();'
         * otherwise snapshot will not write
         */
        writeFileSync = jest
            .spyOn(fs, 'writeFileSync')
            .mockImplementation(() => jest.fn());
    });

    afterEach(() => {
        process.chdir(cwd);
    });

    test('writes stats file', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/stats1/');
        process.chdir(dir);

        const sections = {
            fileManager: {
                directories: [
                    path.normalize('dist'),
                    path.normalize('nested/one'),
                    path.normalize('src'),
                ],
                files: {
                    [path.normalize(
                        '.eslintrc.js',
                    )]: '059a0b7f26dc50ebe483a7eee4534bfd08d0c5816113e3b0b10782e8ab57ec57',
                    [path.normalize(
                        'nested/file1.js',
                    )]: '8b6fd065dc80e6386779fb0188ecfc77fcf4f00fc7f2be965b3035d1577f6347',
                    [path.normalize(
                        'nested/one/file2.js',
                    )]: '6e1878e518fe0e6d4d5445ec84694c99df98b1503eb5c5e91bbd70d0c21114bd',
                },
            },
            packageJson: {
                scripts: {
                    clean: 'backtrack clean',
                    dev: 'backtrack dev --development',
                },
            },
        };

        await writeStatsFile(sections);

        const writeFileMock = writeFileSync.mock.calls;

        writeFileSync.mockRestore();

        const fileContents = JSON.parse(writeFileMock[0][1]);

        expect(fileContents).toEqual({
            fileManager: {
                directories: ['dist', 'nested/one', 'src'],
                files: {
                    '.eslintrc.js':
                        '059a0b7f26dc50ebe483a7eee4534bfd08d0c5816113e3b0b10782e8ab57ec57',
                    'nested/file1.js':
                        '8b6fd065dc80e6386779fb0188ecfc77fcf4f00fc7f2be965b3035d1577f6347',
                    'nested/one/file2.js':
                        '6e1878e518fe0e6d4d5445ec84694c99df98b1503eb5c5e91bbd70d0c21114bd',
                },
            },
            packageJson: {
                scripts: {
                    clean: 'backtrack clean',
                    dev: 'backtrack dev --development',
                },
            },
        });
        expect(writeFileMock).toMatchSnapshot();
        expect(del.mock.calls).toEqual([]);
    });

    test('does not write file if stats have not changed', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/stats1/');
        process.chdir(dir);

        const sections = {
            fileManager: {
                directories: ['src', 'dist'],
                files: {
                    '.eslintrc.js':
                        '059a0b7f26dc50ebe483a7eee4534bfd08d0c5816113e3b0b10782e8ab57ec57',
                },
            },
            packageJson: {
                scripts: {
                    dev: 'backtrack dev --development',
                },
            },
        };

        const previousStats = {
            fileManager: {
                directories: ['src', 'dist'],
                files: {
                    '.eslintrc.js':
                        '059a0b7f26dc50ebe483a7eee4534bfd08d0c5816113e3b0b10782e8ab57ec57',
                },
            },
            packageJson: {
                scripts: {
                    dev: 'backtrack dev --development',
                },
            },
        };

        await writeStatsFile(sections, previousStats);

        const readFileMock = writeFileSync.mock.calls;

        writeFileSync.mockRestore();

        expect(readFileMock).toEqual([]);
        expect(del.mock.calls).toEqual([]);
    });

    test('removes stats file when empty', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/stats1/');
        process.chdir(dir);

        await writeStatsFile({});

        const readFileMock = writeFileSync.mock.calls;

        writeFileSync.mockRestore();

        expect(readFileMock).toEqual([]);
        expect(del.mock.calls).toEqual([
            [path.resolve(dir, '.backtrack-stats.json')],
        ]);
    });

    test('when all sub keys are empty', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/stats1/');
        process.chdir(dir);

        await writeStatsFile({
            fileManager: {},
            packageJson: {},
        });

        const readFileMock = writeFileSync.mock.calls;

        writeFileSync.mockRestore();

        expect(readFileMock).toEqual([]);
        expect(del.mock.calls).toEqual([
            [path.resolve(dir, '.backtrack-stats.json')],
        ]);
    });

    test('removes empty sections', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/stats1/');
        process.chdir(dir);

        await writeStatsFile({
            fileManager: {},
            packageJson: {
                scripts: {
                    dev: 'backtrack dev --development',
                },
            },
        });

        const readFileMock = writeFileSync.mock.calls;

        writeFileSync.mockRestore();

        expect(readFileMock).toMatchSnapshot();
        expect(del.mock.calls).toEqual([]);
    });

    test('do nothing if empty and does not exist', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/stats-missing-file/');
        process.chdir(dir);

        await writeStatsFile({});

        const readFileMock = writeFileSync.mock.calls;

        writeFileSync.mockRestore();

        expect(readFileMock).toEqual([]);
        expect(del.mock.calls).toEqual([]);
    });
});
