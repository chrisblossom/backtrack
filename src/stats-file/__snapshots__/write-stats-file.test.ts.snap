// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`writeStatsFile removes empty sections 1`] = `
Array [
  Array [
    "<PROJECT_ROOT>/.backtrack-stats.json",
    "{
    \\"packageJson\\": {
        \\"scripts\\": {
            \\"dev\\": \\"backtrack dev --development\\"
        }
    }
}
",
  ],
]
`;

exports[`writeStatsFile writes stats file 1`] = `
Array [
  Array [
    "<PROJECT_ROOT>/.backtrack-stats.json",
    "{
    \\"fileManager\\": {
        \\"directories\\": [
            \\"dist\\",
            \\"nested/one\\",
            \\"src\\"
        ],
        \\"files\\": {
            \\".eslintrc.js\\": \\"059a0b7f26dc50ebe483a7eee4534bfd08d0c5816113e3b0b10782e8ab57ec57\\",
            \\"nested/file1.js\\": \\"8b6fd065dc80e6386779fb0188ecfc77fcf4f00fc7f2be965b3035d1577f6347\\",
            \\"nested/one/file2.js\\": \\"6e1878e518fe0e6d4d5445ec84694c99df98b1503eb5c5e91bbd70d0c21114bd\\"
        }
    },
    \\"packageJson\\": {
        \\"scripts\\": {
            \\"clean\\": \\"backtrack clean\\",
            \\"dev\\": \\"backtrack dev --development\\"
        }
    }
}
",
  ],
]
`;
